File: LaunchNetworking/NetworkManager.m
//
//  NetworkManager.m
//  Launches
//
//  Created by Andrew Boissonnault on 12/21/19.
//  Copyright © 2019 Andrew Boissonnault. All rights reserved.
//

#import "NetworkManager.h"
#import <LaunchNetworking/LaunchNetworking-Swift.h>

const NSUInteger NEXT_COUNT = 20;

NS_ASSUME_NONNULL_BEGIN

@interface NetworkManager()

@property (nonatomic) URLBuilder* urlBuilder;
@property (nonatomic) id<URLSessionProtocol> urlSession;

@end


@implementation NetworkManager

-(id)initWithUrlBuilder:(URLBuilder*)urlBuilder urlSession:(id<URLSessionProtocol>)urlSession {
    self = [super init];
    self.urlBuilder = urlBuilder;
    self.urlSession = urlSession;
    return self;
}

-(id)init {
    URLBuilder* builder = [[URLBuilder alloc] init];
    NSURLSession* session = NSURLSession.sharedSession;
    self = [self initWithUrlBuilder:builder urlSession:session];
    return self;
}

-(void)fetchLaunches:(void(^)(NSData *_Nullable data, NSError *_Nullable error))completion {
    NSURL* url = [self.urlBuilder launchUrlWithCount:NEXT_COUNT];
    id<URLSessionDataTaskProtocol> task = [self.urlSession dataTaskWith:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if(error != nil) {
            completion(nil, error);
            return;
        }
        if(data == nil) {
            NSError* error = [[NSError alloc] initWithDomain:LaunchNetworkingErrorDomain code:LaunchNetworkingErrorMissingData userInfo:nil];
            completion(nil, error);
            return;
        }
        completion(data, nil);
    }];
    [task resume];
}

@end

NS_ASSUME_NONNULL_END

File: Launches/Networking/NetworkManager.swift
//
//  NetworkManager.swift
//  Launches
//
//  Created by Andrew Boissonnault on 12/19/19.
//  Copyright © 2019 Andrew Boissonnault. All rights reserved.
//

import Foundation

enum LaunchNetworkingError : Error {
    case missingData
    case jsonSerialization
    case casting
    case networkingError
}

typealias DataTaskResult = (Data?, URLResponse?, Error?) -> Void

protocol URLSessionProtocol {
    func dataTask(with url: URL, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTaskProtocol
}

extension URLSession: URLSessionProtocol {
    func dataTask(with url: URL, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTaskProtocol {
        return (dataTask(with: url, completionHandler: completionHandler)
            as URLSessionDataTask) as URLSessionDataTaskProtocol
    }
}

protocol URLSessionDataTaskProtocol {
    func resume()
}

extension URLSessionDataTask: URLSessionDataTaskProtocol { }

class NetworkManager {
    private let urlBuilder : URLBuilder
    private let urlSession : URLSessionProtocol
    
    init(urlBuilder : URLBuilder, urlSession: URLSessionProtocol) {
        self.urlBuilder = urlBuilder
        self.urlSession = urlSession
    }
    
    convenience init() {
        self.init(urlBuilder: URLBuilder(), urlSession: URLSession.shared)
    }
    
    func fetchLaunches(_ completion: @escaping (_ result: [String:Any]?, _ error: Error?)->()) {
        let url = self.urlBuilder.launchUrl(count: 20)
        let task = self.urlSession.dataTask(with: url) { (data, response, error) in
            if let safeError = error {
                completion(nil, safeError)
                return
            }
            guard let safeData = data else {
                completion(nil, LaunchNetworkingError.missingData)
                return
            }
            guard let json = try? JSONSerialization.jsonObject(with: safeData, options: []) else {
                completion(nil, LaunchNetworkingError.jsonSerialization)
                return
            }
            guard let dictionary = json as? [String : Any] else {
                completion(nil, LaunchNetworkingError.casting)
                return
            }
            completion(dictionary, nil)
        }
        task.resume()
    }
}
